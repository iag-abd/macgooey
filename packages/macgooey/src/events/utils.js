// @flow

import { KEYS } from "./keys";

export const noPropagate = (event: SyntheticEvent<HTMLElement>) => {
  event.stopPropagation();
  return event;
};

export const noImmediatePropagate = (event: SyntheticEvent<HTMLElement>) => {
  event.nativeEvent.stopImmediatePropagation();
  return event;
};

export const preventDefault = (event: SyntheticEvent<HTMLElement>) => {
  event.preventDefault();
  return event;
};

export function keyEvent(keys: $Keys<KEYS>[], handler: mixed => mixed) {
  return (event: SyntheticKeyboardEvent<EventTarget>) => {
    if (keys.includes(event.keyCode)) {
      handler(event);
    }
    return event;
  };
}
