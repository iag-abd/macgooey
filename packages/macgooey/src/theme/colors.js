// @flow

import { type ColorPalette, type ColorShade, type Styles } from "../theme";

type ThemeUtil = {|
  theme: Styles
|};

type ColorUtil = {
  ...ThemeUtil,
  color: string
};

// TODO remove the `Object` type declarations
export const themeColor = (color: $Keys<ColorPalette> = "primary") => ({
  theme
}: ThemeUtil) => theme.color[color];

export const themeColorCompliment = (
  compliment: $Keys<ColorPalette> = "primary"
) => ({ theme }: ThemeUtil) => theme.colorCompliment[compliment];

export const themeColorShade = (shade: $Keys<ColorShade> = "gray") => ({
  theme
}: ThemeUtil) => theme.colorShade[shade];

// If the element takes a `color` prop
export const colorOr = (defaultColor: string = "#000") => ({
  color,
  theme
}: ColorUtil) => theme.color[color] || defaultColor;

export const complimentOr = (defaultColor: string = "#fff") => ({
  color,
  theme
}: ColorUtil) => theme.colorCompliment[color] || defaultColor;

export const colorOrShade = (shade: $Keys<ColorShade> = "lightGray") => ({
  color,
  theme
}: ColorUtil) => theme.color[color] || themeColorShade(shade)({ theme });

export const complimentOrShade = (shade: $Keys<ColorShade> = "lightGray") => ({
  color,
  theme
}: ColorUtil) =>
  theme.colorCompliment[color] || themeColorShade(shade)({ theme });
