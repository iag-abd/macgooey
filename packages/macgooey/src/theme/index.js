// @flow

export { ThemeProvider } from "styled-components";

export type ColorPalette = {|
  primary: string,
  secondary: string,
  tertiary: string
|};

export type ColorShade = {|
  lighterGray: string,
  lightGray: string,
  gray: string,
  darkGray: string,
  darkerGray: string
|};

type ElementStyles = {|
  borderRadius: string,
  borderWidth: string
|};

type ButtonStyles = ElementStyles;

type InputStyles = ElementStyles;

export type Styles = {|
  button: ButtonStyles,
  color: ColorPalette,
  colorCompliment: ColorPalette,
  colorShade: ColorShade,
  fontFamilyHeading: string,
  input: InputStyles,
  textColorBody: string
|};

export const themeBase: Styles = {
  fontFamilyHeading: "'Open Sans', sans-serif",
  textColorBody: "#333",

  button: {
    borderRadius: "3px",
    borderWidth: "0"
  },

  input: {
    borderWidth: "1px",
    borderRadius: "0"
  },

  color: {
    primary: "#4375E8",
    secondary: "#FF4448",
    tertiary: "#E8B221"
  },
  colorCompliment: {
    primary: "#FFF",
    secondary: "#FFF",
    tertiary: "#FFF"
  },
  colorShade: {
    lighterGray: "#eee",
    lightGray: "#ddd",
    gray: "#aaa",
    darkGray: "#555",
    darkerGray: "#222"
  }
};

export const themeAerith = {
  ...themeBase,

  fontFamilyHeading: "'Montserrat', sans-serif",
  textColorBody: "#414042",

  input: {
    borderWidth: "1px",
    borderRadius: "0"
  },

  color: {
    default: "#DDD",
    primary: "#0BAB47",
    secondary: "#6D6E71",
    tertiary: "#007BAD"
  },
  colorCompliment: {
    default: "#333",
    primary: "#FFF",
    secondary: "#FFF",
    tertiary: "#FFF"
  }
};

export const themeSephiroth = {
  ...themeBase,

  fontFamilyHeading: "'Source Sans Pro', 'Open Sans', sans-serif",
  textColorBody: "#3B3D40",

  button: {
    borderRadius: "3px",
    borderWidth: "1px"
  },

  color: {
    default: "#F0F0F0",
    primary: "#602f89",
    secondary: "#01B5CC",
    tertiary: "#1E2741"
  },
  colorCompliment: {
    default: "#602F89",
    primary: "#FFF",
    secondary: "#FFF",
    tertiary: "#FFF"
  }
};
