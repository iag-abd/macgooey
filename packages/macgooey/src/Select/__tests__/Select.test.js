// @flow

import React from "react";
import { Select } from "..";
import { themeBase as theme } from "../../theme";
import { mountStyled, withBaseTheme } from "../../theme/__tests__/utils";
import "jest-styled-components";

import renderer from "react-test-renderer";
import { shallow, mount } from "enzyme";

describe("<Select />", () => {
  test("it should render", () => {
    const component = renderer.create(<Select options={[]} theme={theme} />);

    expect(component.toJSON()).toMatchSnapshot();
  });
});

describe("<Select options />", () => {
  const testOptions = [{ text: "foo", value: 1 }, { text: "bar", value: 2 }];

  test("it should open after a click event on the closed select input", () => {
    const component = shallow(<Select options={testOptions} theme={theme} />);
    component.simulate("click");

    expect(component.find("OptionList").length).toEqual(1);
  });
  test("it should NOT open after a keydown of the Right Arrow when the select input has focus", () => {
    const component = shallow(<Select options={testOptions} theme={theme} />);
    component.simulate("keydown", { keyCode: 39 });

    expect(component.find("OptionList").length).toEqual(0);
  });
  test("it should open after a keydown of the Down Arrow when the select input has focus", () => {
    const component = shallow(<Select options={testOptions} theme={theme} />);
    component.simulate("keydown", { keyCode: 40 });

    expect(component.find("OptionList").length).toEqual(1);
  });
  test("it should open after a keydown of the Up Arrow when the select input has focus", () => {
    const component = shallow(<Select options={testOptions} theme={theme} />);
    component.simulate("keydown", { keyCode: 38 });

    expect(component.find("OptionList").length).toEqual(1);
  });

  test("it should render all of the options in the list", () => {
    const component = mount(withBaseTheme(<Select options={testOptions} />));
    component.simulate("keydown", { keyCode: 38 });
    const options = component.find("Option");

    expect(options.length).toEqual(2);
  });

  test("it should focus on the first option when Arrow Down key is pressed a second time", () => {
    const component = mount(withBaseTheme(<Select options={testOptions} />));

    component.simulate("keydown", { keyCode: 40 });
    component.simulate("keydown", { keyCode: 40 });

    const options = component.find("Option");

    expect(options.first().prop("data-focused")).toEqual(true);
  });

  test("it should close the select when an option is selected and the select text should update", () => {
    const component = mount(withBaseTheme(<Select options={testOptions} />));

    component.simulate("click");
    component.find("Option").at(0).simulate("mouseup");

    expect(component.find("Option").length).toEqual(0);
    expect(component.find("span").text()).toEqual("foo");
  });

  test("it should scroll the element into view when the select is opened with a value set", () => {
    const component = mount(withBaseTheme(<Select defaultValue={1} options={testOptions} />));

    // We'll mock this temporarily
    const sivMock = jest.fn();
    window.HTMLElement.prototype.scrollIntoView = sivMock;

    component.simulate("click");

    expect(sivMock.mock.calls.length).toBe(1);

    // undo mocking in case we screw something else up
    delete window.HTMLElement.prototype.scrollIntoView;
  });
});

describe("<Select placeholder />", () => {
  const testOptions = [{ text: "foo", value: 1 }, { text: "bar", value: 2 }];

  test("it should display the default placeholder when nothing is selected", () => {
    const component = renderer.create(
      <Select options={testOptions} theme={theme} />
    );

    expect(component.toJSON()).toMatchSnapshot();
  });

  test("it should display the custom placeholder when nothing is selected", () => {
    const component = renderer.create(
      <Select
        placeholder="Foo Placeholder"
        options={testOptions}
        theme={theme}
      />
    );

    expect(component.toJSON()).toMatchSnapshot();
  });
});
