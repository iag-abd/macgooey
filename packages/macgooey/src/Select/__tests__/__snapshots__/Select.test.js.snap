// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Select /> it should render 1`] = `
.c0 {
  padding: 10px;
  font-family: 'Open Sans',sans-serif;
  font-size: 1em;
  border-width: 1px;
  border-color: #ddd;
  border-style: solid;
  outline: none;
  width: 100%;
  -webkit-transition: border-color 150ms;
  transition: border-color 150ms;
  background: #eee;
  position: relative;
  -webkit-user-select: none;
  -moz-user-select: none;
  -ms-user-select: none;
  user-select: none;
}

.c0:focus {
  border-color: #4375E8;
}

.c0[data-selected="false"] {
  color: #aaa;
}

<div
  className="c0"
  data-selected={false}
  onClick={[Function]}
  onKeyDown={[Function]}
  onMouseDown={[Function]}
  tabIndex={0}
>
  <span>
    Please select
  </span>
</div>
`;
